fix(client): add tooltip for calendar-heatmap

moT01 opened the pull request:
This fixes up the heatmap calendar a little.
- Add's the tooltip back
- added react-tooltip package for the tooltip
- rework the function to create the data for the map

<!-- -->
- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

I'm hoping it can close these issues:
https://github.com/freeCodeCamp/freeCodeCamp/issues/35916
https://github.com/freeCodeCamp/freeCodeCamp/issues/17299
https://github.com/freeCodeCamp/freeCodeCamp/issues/17822
https://github.com/freeCodeCamp/freeCodeCamp/issues/22031

moT01 committed:
fix/calendar-heatmap

moT01 committed:
fix/remove-console-logs

moT01 committed:
fix/conditionals-for-items-completed-on-tooltip

moT01 committed:
fix/remove-gutter

thecodingaviator commented:
I guess you'll need to remove `package-lock.json` ?

moT01 commented:
I'm not sure - I thought it should be kept on when adding a package - did you get a chance to test it out @thecodingaviator?

thecodingaviator commented:
Yes, on doing `npm i`, I get this:

```bash
C:\Users\hp\Desktop\Web\freeCodeCamp>npm i

> @freecodecamp/freecodecamp@0.0.1 postinstall C:\Users\hp\Desktop\Web\freeCodeCamp
> npm run bootstrap


> @freecodecamp/freecodecamp@0.0.1 bootstrap C:\Users\hp\Desktop\Web\freeCodeCamp
> lerna bootstrap --ci

lerna notice cli v3.13.1
lerna info versioning independent
lerna info ci enabled
lerna info Bootstrapping 9 packages
lerna info Installing external dependencies
lerna ERR! npm ci exited 4294963248 in '@freecodecamp/client'
lerna ERR! npm ci stderr:
npm WARN prepare removing existing node_modules/ before installation
WARN tarball tarball data for react-ga@2.5.7 (sha512-UmATFaZpEQDO96KFjB5FRLcT6hFcwaxOmAJZnjrSiFN/msTqylq9G+z5Z8TYzN/dbamDTiWf92m6MnXXJkAivQ==) seems to be corrupted. Trying one more time.
npm ERR! path C:\Users\hp\Desktop\Web\freeCodeCamp\client\node_modules\react-ga\dist\react-ga.js
npm ERR! code EPERM
npm ERR! errno -4048
npm ERR! syscall unlink
npm ERR! Error: EPERM: operation not permitted, unlink 'C:\Users\hp\Desktop\Web\freeCodeCamp\client\node_modules\react-ga\dist\react-ga.js'
npm ERR!  { [Error: EPERM: operation not permitted, unlink 'C:\Users\hp\Desktop\Web\freeCodeCamp\client\node_modules\react-ga\dist\react-ga.js']
npm ERR!   cause:
npm ERR!    { Error: EPERM: operation not permitted, unlink 'C:\Users\hp\Desktop\Web\freeCodeCamp\client\node_modules\react-ga\dist\react-ga.js'
npm ERR!      type: 'OperationalError',
npm ERR!      '$error': '$error',
npm ERR!      cause:
npm ERR!       { errno: -4048,
npm ERR!         code: 'EPERM',
npm ERR!         syscall: 'unlink',
npm ERR!         path:
npm ERR!          'C:\\Users\\hp\\Desktop\\Web\\freeCodeCamp\\client\\node_modules\\react-ga\\dist\\react-ga.js' },
npm ERR!      isOperational: true,
npm ERR!      errno: -4048,
npm ERR!      code: 'EPERM',
npm ERR!      syscall: 'unlink',
npm ERR!      path:
npm ERR!       'C:\\Users\\hp\\Desktop\\Web\\freeCodeCamp\\client\\node_modules\\react-ga\\dist\\react-ga.js' },
npm ERR!   isOperational: true,
npm ERR!   stack:
npm ERR!    'Error: EPERM: operation not permitted, unlink \'C:\\Users\\hp\\Desktop\\Web\\freeCodeCamp\\client\\node_modules\\react-ga\\dist\\react-ga.js\'',
npm ERR!   type: 'OperationalError',
npm ERR!   '$error': '$error',
npm ERR!   errno: -4048,
npm ERR!   code: 'EPERM',
npm ERR!   syscall: 'unlink',
npm ERR!   path:
npm ERR!    'C:\\Users\\hp\\Desktop\\Web\\freeCodeCamp\\client\\node_modules\\react-ga\\dist\\react-ga.js' }
npm ERR!
npm ERR! The operation was rejected by your operating system.
npm ERR! It's possible that the file was already in use (by a text editor or antivirus),
npm ERR! or that you lack permissions to access it.
npm ERR!
npm ERR! If you believe this might be a permissions issue, please double-check the
npm ERR! permissions of the file and its containing directories, or try running
npm ERR! the command again as root/Administrator (though this is not recommended).

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\hp\AppData\Roaming\npm-cache\_logs\2019-05-14T08_06_08_395Z-debug.log

lerna ERR! npm ci exited 4294963248 in '@freecodecamp/client'
npm ERR! code ELIFECYCLE
npm ERR! errno 4294963248
npm ERR! @freecodecamp/freecodecamp@0.0.1 bootstrap: `lerna bootstrap --ci`
npm ERR! Exit status 4294963248
npm ERR!
npm ERR! Failed at the @freecodecamp/freecodecamp@0.0.1 bootstrap script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\hp\AppData\Roaming\npm-cache\_logs\2019-05-14T08_06_09_758Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 4294963248
npm ERR! @freecodecamp/freecodecamp@0.0.1 postinstall: `npm run bootstrap`
npm ERR! Exit status 4294963248
npm ERR!
npm ERR! Failed at the @freecodecamp/freecodecamp@0.0.1 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\hp\AppData\Roaming\npm-cache\_logs\2019-05-14T08_06_09_975Z-debug.log
```

moT01 commented:
so you're thinking that's from the package-lock file?

ojeytonwilliams commented:
@moT01 I'm taking a look at this now.  Is there an easy way to recreate those issues locally?

thecodingaviator commented:
> so you're thinking that's from the package-lock file?

I'm not sure why else would it be happening, I'll try deleting the file and retrying once

moT01 commented:
Honestly, no @ojeytonwilliams - I couldn't recreate most of the issues anywhere - so, I'm not sure what to do with it. I added the tooltip back in, which must have got removed at some point - and reworked the logic that creates the map to hopefully make it a little more robust.

ojeytonwilliams commented:
@moT01 Understood.  I think the issues are all with a very old version of this, anyway, so some might have been fixed in the move to `react-calendar-heatmap`.

moT01 commented:
To be a little bit more clear - when I added this tooltip in - the `calendarValues` variable only included data for dates which had activity from the user - and the tooltip would show like "null" for the items and date with no activity - so I added an empty object and filled it with all the dates starting from six months ago to give the tooltip some info for those dates. And the logic got reworked mostly because I needed that.

ojeytonwilliams commented on client/src/components/profile/components/HeatMap.js:
```diff
@@ -57,18 +68,30 @@ function HeatMap({ calendar, streak }) {
         <CalendarHeatMap
           classForValue={value => {
-            if (!value) {
+            if (value.count < 1) {
```
```suggestion
            if (!value || value.count < 1) {
```
Just in case value somehow doesn't get set.

ojeytonwilliams commented on client/src/components/profile/components/HeatMap.js:
```diff
@@ -57,18 +68,30 @@ function HeatMap({ calendar, streak }) {
+          startDate={startOfCalendar}
+          tooltipDataAttrs={value => {
+            let valueCount = '';
+            if (value.count === 1) {
```
```suggestion
            if (value && value.count === 1) {
```
More defensive coding

ojeytonwilliams commented on client/src/components/profile/components/HeatMap.js:
```diff
@@ -57,18 +68,30 @@ function HeatMap({ calendar, streak }) {
+            let valueCount = '';
+            if (value.count === 1) {
+              valueCount = `${value.count} item on `;
+            } else if (value.count > 1) {
```
```suggestion
            } else if (value && value.count > 1) {
```
and again

ojeytonwilliams commented on client/src/components/profile/components/HeatMap.js:
```diff
@@ -2,6 +2,9 @@ import React from 'react';
+import ReactTooltip from 'react-tooltip';
+import addDays from 'date-fns/add_days';
+import addMonths from 'date-fns/add_months';
 import startOfMonth from 'date-fns/start_of_month';
```
```suggestion
```
If you do go with my suggestion about the start, this can go.  Otherwise, just ignore this comment!

ojeytonwilliams reviewed (changes requested):
This is a nice PR.  While I ended up writing a lot of comments, they're mostly me worrying about edge cases that may never turn up.  Overall I think this is a nice change.

Regarding the start of the calendar, I just want to stress that it's not something I'd hold up the merge for.  I just wanted to see what you thought about it.

moT01 committed:
fix/suggested-changes

moT01 commented:
Yes, I like those suggestions @ojeytonwilliams - I obviously started at the beginning of six months ago because that's how it was, maybe there was a reason for that. But it seems to make more sense, to me anyway, doing it how you suggested - so I made those changes

ojeytonwilliams commented:
@moT01 Perhaps so - this looks nice though.  Thanks for making those changes, it LGTM now.

@ahmadabdolsaheb Does everything look okay to you?

ValeraS commented on client/src/components/profile/components/HeatMap.js:
```diff
@@ -20,29 +22,32 @@ const propTypes = {
-  const calendarValues = Object.keys(dateValueMap).map(key => ({
+  for (let timestamp in calendar) {
+    if (calendar.hasOwnProperty(timestamp)) {
+      timestamp = Number(timestamp * 1000) || null;
```
Why do not use the old code and change only the last line?

```js
const dateValueMap = Object.keys(calendar)
    .map(ts => Number(ts * 1000) || null)
    .filter(Boolean)
    .reduce((map, current) => {
      const startOfCurrent = format(startOfDay(current), 'YYYY-MM-DD');
      if (startOfCurrent in map) {
        map[startOfCurrent] = map[startOfCurrent] + 1;
      } else {
        map[startOfCurrent] = 1;
      }
      return map;
    }, calendarData);
```

moT01 replied:
I don't know, I liked it better - it's shorter

ValeraS replied:
What happens if `timestamp` is null?

moT01 replied:
I guess it all blows up... it really shouldn't ever be null - but yea, you can't count on that, I suppose.
Could add an `if` in there maybe... ?
```
if(timestamp) {
      const startOfTimestampDay = format(startOfDay(timestamp), 'YYYY-MM-DD');
      calendarData[startOfTimestampDay] =
      calendarData[startOfTimestampDay] + 1 || 1;
}
```

ojeytonwilliams replied:
Turns out they get added to 1970-01-01 (i.e. `null` is treated like 0).  This is out of the range specified by `endOfCalendar` and `startOfCalendar`, so doesn't get displayed.

That said, probably best if they never get added to `calendarData` so adding the `if` seems reasonable to me.

moT01 replied:
I added the condition. Let me know if we want something else.

moT01 committed:
fix/add-condition-for-timestamp

ValeraS committed:
fix(client): remove unused parts of old heatmap and adjust a tooltip

ValeraS reviewed (approved):
@moT01 if you agree with changes I will merge it.

moT01 commented on client/src/components/profile/components/HeatMap.js:
```diff
@@ -75,19 +67,27 @@ function HeatMap({ calendar, streak }) {
+              valueCount = 'No items';
             }
             return {
-              'data-tip': `${valueCount}${format(value.date, 'MMMM Do, YYYY')}`
```
I'm curious why you changed this @ValeraS?

moT01 reviewed (commented):
Yes, I'm fine with these changes @ValeraS, but I had a thought... maybe we should change all the "item/items" to "point/points" - since that is how it's described above the heatmap (see picture) - also, I noticed that the "current streak" defaults to `1` - since it's right there in this file, maybe we should change that to default to zero in this PR.
![Screen Shot 2019-05-26 at 6 34 51 PM](https://user-images.githubusercontent.com/20648924/58388509-8fe68b80-7fe5-11e9-810b-04d374d9332d.png)

ValeraS commented:
> Yes, I'm fine with these changes @ValeraS, but I had a thought... maybe we should change all the "item/items" to "point/points" - since that is how it's described above the heatmap (see picture) - also, I noticed that the "current streak" defaults to `1` - since it's right there in this file, maybe we should change that to default to zero in this PR.

Let's rename to `point/points` and change defaults for both streaks to `0`.

moT01 committed:
fix/current-streak+longest-streak-default-to-0

ValeraS committed:
fix: use fixed string for 1 point and en-US locale for date
